plugins/out_s3/s3.c

```diff
@@ -1343,6 +1348,137 @@ static void cb_s3_upload(struct flb_config *config, void *data)
     }
 }
 
+static flb_sds_t flb_pack_msgpack_extract_log_key(void *out_context, const char *data,
+                                                  uint64_t bytes)
+{
+    int i;
+    int records = 0;
+    int map_size;
+    int check = FLB_FALSE;
+    int found = FLB_FALSE;
+    int log_key_missing = FLB_FALSE;
+    int ret;
+    struct flb_s3 *ctx = out_context;
+    char *val_buf;
+    char *orig_val_buf;
+    char *key_str = NULL;
+    size_t key_str_size = 0;
+    size_t off = 0;
+    flb_sds_t out_buf = NULL;
+    msgpack_unpacked result;
+    msgpack_object root;
+    msgpack_object map;
+    msgpack_sbuffer tmp_sbuf;
+    msgpack_object key;
+    msgpack_object val;
+
+    /* Iterate the original buffer and perform adjustments */
+    records = flb_mp_count(data, bytes);
+    if (records <= 0) {
+        return NULL;
+    }
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&tmp_sbuf);
+
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        /* Each array must have two entries: time and record */
+        root = result.data;
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            continue;
+        }
+        if (root.via.array.size != 2) {
+            continue;
+        }
+
+        /* Get the record/map */
+        map = root.via.array.ptr[1];
+        if (map.type != MSGPACK_OBJECT_MAP) {
+            continue;
+        }
+        map_size = map.via.map.size;
+
+        /* Extract log_key from record and append to output buffer */
+        for (i = 0; i < map_size; i++) {
+            key = map.via.map.ptr[i].key;
+            val = map.via.map.ptr[i].val;
+
+            if (key.type == MSGPACK_OBJECT_BIN) {
+                key_str  = (char *) key.via.bin.ptr;
+                key_str_size = key.via.bin.size;
+                check = FLB_TRUE;
+            }
+            if (key.type == MSGPACK_OBJECT_STR) {
+                key_str  = (char *) key.via.str.ptr;
+                key_str_size = key.via.str.size;
+                check = FLB_TRUE;
+            }
+
+            if (check == FLB_TRUE) {
+                if (strncmp(ctx->log_key, key_str, key_str_size) == 0) {
+                    found = FLB_TRUE;
+
+                    orig_val_buf = flb_malloc(bytes + bytes / 4);
```

Comment:
allocated 1.25 * bytes here doesn't make sense to me. bytes is the entire size of the msgpack, but this is inside an iteration of the loop, where you are only dealing with a single record.

If possible write this code so that you do not allocate any memory inside the loop. that will make it much more efficient and faster. One thing with C code is that allocating a lot of memory should often be considered to be completely fine- the machines fluent bit is run on usually have a lot of memory. For efficiency, you should often care more about keeping those allocs infrequent, because calls to alloc memory are much slower than any other computation. 

And for this code, I see no reason why you can't alloc one large buffer outside the loop, and then use it within the loop. You might possibly need multiple buffers, some for copying and temporarily storing data. But the point is to try to see if you can do everything by alloc'ing a few buffers before/after the loop. Nothing inside of it.